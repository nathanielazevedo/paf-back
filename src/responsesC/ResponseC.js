/** @format */

import "./ResponseC.css";
import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import Form from "../form/Form";
import Paf from "../api.js";
import { Button } from "reactstrap";
import Response from "../response/Response";

//All api requests regarding responses are initiated here. (CRUD)

function ResponseC() {
  const { id } = useParams();
  const [form, setForm] = useState(false);
  const [responses, setResponses] = useState([]);
  const [statement, setStatement] = useState();

  const openForm = () => {
    setForm(true);
  };
  const closeForm = () => {
    setForm(false);
  };

  //Get all responses upon render
  useEffect(
    function () {
      async function getStatementInfo() {
        let statementy = await Paf.getStatementInfo(id);
        setResponses(statementy.responses);
        delete statementy.responses;
        setStatement(statementy.statement);
      }
      getStatementInfo();
    },
    [id]
  );

  //ADD RESPONSE
  const addResponse = async (formData) => {
    formData.statement_id = parseInt(id);
    let res = await Paf.addStatementResponse(formData);
    setResponses((data) => {
      if (data) {
        return [...data, res];
      } else {
        return [res];
      }
    });
    setForm(false);
    return { success: true };
  };

  // Edit RESPONSE
  async function editResponse(id, data) {
    await Paf.editResponse(id, data);
    let pos = responses
      .map((s) => {
        return s.id;
      })
      .indexOf(id);

    setResponses((datas) => {
      datas[pos].response = data.response;

      return [...datas];
    });
    return { success: true };
  }

  // Delete a response
  async function deleteResponse(id) {
    await Paf.deleteResponse(id);
    let pos = responses
      .map((s) => {
        return s.id;
      })
      .indexOf(id);

    setResponses((data) => {
      data.splice(pos, 1);

      return [...data];
    });
  }

  return (
    <>
      <ul className="friend-convo-list">
        <li className="friend-i-say title">
          When I say {statement ? statement : null}
        </li>
        <li className="friend-i-say title">My friend can say...</li>
        {responses
          ? responses.map((s, i) => {
              return (
                <Response
                  s={s}
                  deleteFunc={deleteResponse}
                  editFunc={editResponse}
                  key={i}
                />
              );
            })
          : null}
        <li className="container-button">
          <Button onClick={openForm}>Add Response</Button>
        </li>
      </ul>
      {form ? (
        <Form
          title="New Response"
          inputs={["response"]}
          func={addResponse}
          after={`/paf-front-end/responses/${id}`}
          close={closeForm}
        />
      ) : null}
    </>
  );
}

export default ResponseC;
