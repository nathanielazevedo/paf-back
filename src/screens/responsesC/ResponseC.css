/** @format */
/* 
.friend-convo-list {
  width: 100%;
  min-height: 80vh;
  height: fit-content;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: relative;
  margin-bottom: 50px;
  padding-top: 25px;
  padding-bottom: 25px;
} */


.response{
  float: left;
  margin-bottom: 50px;
  align-self: flex-start;
  box-shadow: inset 0 0 0 0 #31302B;
}



