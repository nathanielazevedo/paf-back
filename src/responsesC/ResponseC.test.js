/** @format */

import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import ResponseC from "./ResponseC";
import { UserProvider } from "../testUtils";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <ResponseC />
      </UserProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", function () {

  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <ResponseC/>
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("displays expected text", function () {
  const { getAllByText } = render(
    <MemoryRouter>
      <UserProvider>
        <ResponseC />
      </UserProvider>
    </MemoryRouter>
  );

  expect(getAllByText("Add Response")[0]).toBeInTheDocument();
});  