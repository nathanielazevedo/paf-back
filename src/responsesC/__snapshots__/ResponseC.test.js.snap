// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="friend-convo-list"
  >
    <li
      class="friend-i-say title"
    >
      When I say 
    </li>
    <li
      class="friend-i-say title"
    >
      My friend can say...
    </li>
    <li
      class="container-button"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Add Response
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
